File: 문제풀이/온라인 저지/백준/# 08. Silver Ⅲ/# 02108. 통계학/java/Main.java
import java.io.DataInputStream;
import java.io.IOException;
import java.io.Reader;

public class Main {

    public static void main(String[] args) throws IOException {
        Reader r = new Reader();
        int n = r.nextInt();
        int[] counter = new int[8001];

        int max = -4001;
        int min = 4001;
        int sum = 0;
        int value;
        for (int i=0; i<n; i++) {
            value = r.nextInt();
            counter[value + 4000] ++;
            if (value > max) {
                max = value;
            }

            if (value < min) {
                min = value;
            }
            sum += value;
        }
        int mode = -4001;
        int modeCount = 0;
        boolean modeDuplicated = false;

        int count = 0;
        int median = 4001;
        for (int i=min + 4000; i < max + 4001; i++) {
            if (counter[i] > 0) {
                count += counter[i];
                if (median == 4001 && count > n/2) {
                    median = i - 4000;
                }
                if (counter[i] > modeCount) {
                    mode = i - 4000;
                    modeCount = counter[i];
                    modeDuplicated = false;
                } else if (counter[i] == modeCount && !modeDuplicated){
                    mode = i - 4000;
                    modeDuplicated = true;
                }
            }
        }

        StringBuilder sb = new StringBuilder();
        sb.append(Math.round(sum/(double)n))
                .append('\n')
                .append(median)
                .append('\n')
                .append(mode)
                .append('\n')
                .append(max-min);
        System.out.print(sb);
    }


    static class Reader {
        private final int BUFFER_SIZE = 1 << 16;
        private DataInputStream din;
        private byte[] buffer;
        private int bufferPointer, bytesRead;

        public Reader() {
            din = new DataInputStream(System.in);
            buffer = new byte[BUFFER_SIZE];
            bufferPointer = bytesRead = 0;
        }

        public int nextInt() throws IOException {
            int ret = 0;
            byte c = read();
            while (c <= ' ') {
                c = read();
            }
            boolean neg = (c == '-');
            if (neg)
                c = read();
            do {
                ret = ret * 10 + c - '0';
            } while ((c = read()) >= '0' && c <= '9');

            if (neg)
                return -ret;
            return ret;
        }

        private void fillBuffer() throws IOException {
            bytesRead = din.read(buffer, bufferPointer = 0,
                    BUFFER_SIZE);
            if (bytesRead == -1)
                buffer[0] = -1;
        }

        private byte read() throws IOException {
            if (bufferPointer == bytesRead)
                fillBuffer();
            return buffer[bufferPointer++];
        }

    }
}

File: 문제풀이/온라인 저지/백준/# 08. Silver Ⅲ/# 02108. 통계학/kotlin/Main.kt
import kotlin.math.roundToInt

fun main() {
        val n = i()
        val counter = IntArray(8001)
        var v: Int
        var avg = 0
        var min = 4001
        var max = -4001
        for (i in 1..n) {
            v = i()
            counter[v + 4000] ++
            avg += v
            if (max < v) max = v
            if (min > v) min = v
        }
        avg = (avg/n.toDouble()).roundToInt()


        var cnt = 0
        var mode = -4001
        var modeCnt = 0
        var duplicate = false
        var mid = 4001
        for (i in min+4000..max+4000) {
            if (counter[i] > 0) {
                cnt += counter[i]
                if (mid == 4001 && cnt > n/2) {
                    mid = i - 4000
                }
                if (counter[i] > modeCnt) {
                    mode = i - 4000
                    modeCnt = counter[i]
                    duplicate = false
                } else if (counter[i] == modeCnt && !duplicate) {
                    mode = i - 4000
                    duplicate = true
                }
            }
        }
        val sb = StringBuilder()
        sb.append(avg).append('\n')
            .append(mid).append('\n')
            .append(mode).append('\n')
            .append(max-min)
        print(sb)
    }
    private const val S = 65536
    private val iS = java.io.DataInputStream(System.`in`)
    private val b = ByteArray(S)
    private var c = 0
    private var l = 0

    private fun i(): Int {
        var v = 0
        var c = r()
        val n = c.toInt() == 45
        c = if (n) r() else c
        do { v = v * 10 + c - 48 } while (r().also { c = it } > 47)
        return if (n) -v else v
    }

    private fun r(): Byte {
        if (c == l) {
            l = iS.read(b, 0.also { c = it }, S)
            if (l == -1) b[0] = -1
        }
        return b[c++]
    }